Separate tests that rely on SecurityManager (Java8andUp)

EricYangIBM opened the pull request:
SecurityManager will be removed in a future jdk. Separate tests that
rely on it so that they can be easily disabled in the future.

Issue: #14412
Signed-off-by: Eric Yang <eric.yang@ibm.com>

JasonFengJ9 commented on test/functional/Java8andUp/testng.xml:
```diff
@@ -247,14 +247,19 @@
+		<!-- Tests that depend on SecurityManager -->
+		<classes>
+			<class name="org.openj9.test.java.security.Test_AccessController"/>
+			<class name="org.openj9.test.java.security.Test_AccessControlContext"/>
```
The security manager related class `AccessController` & `AccessControlContext` will be removed in future release, the test compilation need to be separated as well.

EricYangIBM replied:
I thought we would disable compilation later when the classes are actually removed. (Done in build.xml as a new target=compile section? e.g. `<equals arg1="${JDK_VERSION}" arg2="19" />`

JasonFengJ9 replied:
https://github.com/eclipse-openj9/openj9/issues/14412#issuecomment-1027506959 predicted the security manager to be removed around JDK19 release.
I think separating the test compilation/running is preferred which is easier for the migration to a Java release w/o security manger classes.

EricYangIBM replied:
Perhaps we could disable the compilation in a separate PR? That way we can disable compilation only when SecurityManager is removed

JasonFengJ9 replied:
> Perhaps we could disable the compilation in a separate PR?

Yes, disabling the compilation can happen when the classes are removed. 
Separating the compilation of those security manager tests can happen, in one or multiple PRs, which needn't wait until the actual removal.

EricYangIBM replied:
Would I do this by adding a new exclude here with the condition that `${JDK_VERSION}` is >=19? https://github.com/eclipse-openj9/openj9/blob/master/test/functional/Java8andUp/build.xml#L284

JasonFengJ9 replied:
That works w/ `Test_AccessController` & `Test_AccessControlContext`, there are other not dedicated security manager related tests but contain references to `AccessControlContext`, `AccessController`, `SecurityManager`, etc. such as https://github.com/eclipse-openj9/openj9/blob/99077564bf9bd2a32409b5861e8b8687061cddc0/test/functional/Java8andUp/src_110_up/org/openj9/test/java/lang/Test_Class.java#L710-L711
Those code have to be separated from their existing host classes, it is tedious, not sure if there is a better approach.

EricYangIBM commented:
Added compilation excludes

pshipton commented:
We don't want to actually exclude the tests for jdk19+ until it's determined the SecurityManager is removed in jdk19. I was just a guess. We just want to be ready to do so quickly, so we aren't blocked from running tests for a month when the change happens.

JasonFengJ9 commented:
> We don't want to actually exclude the tests for jdk19+ until it's determined the SecurityManager is removed in jdk19. I was just a guess. We want to be ready to do so quickly, so we aren't blocked from running tests for a month when the change happens.

I think this PR is preparing us to continue running tests when the SecurityManger related classes are removed in JDK19 or a future release. If it doesn't happen at JDK19, the change required is to just put JDK19 back to the allowed Java levels `<matches string="${JDK_VERSION}" pattern="^(8|9|(1[0-8]))$$" />` (one line modification), otherwise this PR will have to be developed in short time period along with quite a few other places.

A list of deprecated APIs for removal:
```
java.lang.System::{setSecurityManager, getSecurityManager}
java.security.{Policy, PolicySpi, Policy.Parameters}
java.security.{AccessController, AccessControlContext, AccessControlException, DomainCombiner}

java.lang.Thread::checkAccess
java.lang.ThreadGroup::checkAccess
java.util.logging.LogManager::checkAccess
java.util.concurrent.Executors::{privilegedCallable, privilegedCallableUsingCurrentClassLoader, privilegedThreadFactory}
java.rmi.RMISecurityManager
javax.security.auth.SubjectDomainCombiner and javax.security.auth.Subject::{doAsPrivileged, getSubject} 
```
fyi @EricYangIBM

EricYangIBM commented on test/functional/Java8andUp/playlist.xml:
```diff
@@ -1492,6 +1492,48 @@
+			<version>15</version>
+			<version>16</version>
+			<version>17</version>
+			<version>18</version>
```
We will also have to add to versions. (Is there a better way than listing all the versions individually?)

JasonFengJ9 commented:
Are there more changes coming for following tests?
```
Java8andUp/src/org/openj9/test/annotationPackage/Test_PackageAnnotation.java
Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java
Java8andUp/src_80/org/openj9/test/java/lang/Test_Class.java
Java8andUp/src_90/org/openj9/test/java/lang/Test_Class.java
```

EricYangIBM commented:
>Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java

Is this related to security manager? `-Djava.security.policy=/opt/IBM/WebSphere/AppServer80/profiles/AppSrv01/properties/server.policy`

JasonFengJ9 commented:
> Is this related to security manager? -Djava.security.policy=/opt/IBM/WebSphere/AppServer80/profiles/AppSrv01/properties/server.policy

Usually `java.security.policy` is related to SecurityManager [1].

[1] https://docs.oracle.com/javase/8/docs/technotes/guides/security/PolicyFiles.html

EricYangIBM commented:
For the two `Test_Class.java`, since they are `src_80` and `src_90` won't they not be compiled for jdk19+?

JasonFengJ9 commented:
> For the two Test_Class.java, since they are src_80 and src_90

Scratch these two then, on the other hand, it means `src_90` isn't used at all since there is no `Java 9` build/test.

EricYangIBM commented:
I added the runtime check for VmArgumentTests and removed the `AccessController.doPrivileged` in Test_PackageAnnotation (SecurityException wasn't thrown, also only used for logging so probably fine)

EricYangIBM commented on test/functional/Java8andUp/build.xml:
```diff
@@ -264,33 +264,81 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
+							<exclude name="org/openj9/test/java/lang/invoke/Test_MethodHandleInfo_SM.java" />
+							<exclude name="org/openj9/test/java/lang/Test_ClassLoader_SM.java" />
+							<exclude name="org/openj9/test/java/lang/Test_J9VMInternals.java" />
+							<exclude name="org/openj9/test/java/lang/Test_J9VMInternalsImpl.java" />
```
Based on the comment `TestNG parses the test file, which interferes with the classloading under test` I added the security manager test classes with the same two file pattern

JasonFengJ9 reviewed (commented):
More cleanup for new classes please.

JasonFengJ9 commented:
Requesting reviews from @llxia @pshipton before asking for rebase & squash the commits.

llxia commented:
@renfeiw could you help to review this PR? Thanks

EricYangIBM commented:
I have two other smaller PRs for `JLM_Tests` and `Jsr292` ready: #14562 #14580
@JasonFengJ9 could you please review these as well?

pshipton commented:
This isn't working on vmfarm, see /jobs_by_status.php?build_id=23133&status=FAILED
```
[TestNG] [ERROR] 
Cannot find class in classpath: org.openj9.test.java.lang.Test_Class_SM
Exception in thread "main" java.lang.NullPointerException
	at org.testng.TestNG.getStatus(TestNG.java:211)
	at org.testng.TestNG.main(TestNG.java:1324)
```
I didn't try jenkins (yet) but I assume it's been tested there.

EricYangIBM commented:
I think the problem is that it is looking for `Test_Class_SM` and `Test_System_SM`, both only in `src_110_up` right now. I will need to add duplicates or move them to `src`.
Also I will need to add the new SM tests to SE80 tests

EricYangIBM commented:
Sanity.functional passes for jdk8 and 11: https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/12062/

JasonFengJ9 commented:
JavaNext(19) doesn't work.
```
===============================================
Running test truncatedReturnTest_0 ...
===============================================
truncatedReturnTest_0 Start Time: Fri Feb 25 09:45:22 2022 Epoch Time (ms): 1645811122799
variation: NoOptions
JVM_OPTIONS:  
[TestNG] [ERROR]
Cannot find class in classpath: org.openj9.test.java.lang.Test_ClassLoader_SM
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "org.testng.internal.ExitCodeListener.hasTests()" because "this.exitCodeListener" is null
        at org.testng.TestNG.getStatus(TestNG.java:211)
        at org.testng.TestNG.main(TestNG.java:1324)

truncatedReturnTest_0_FAILED
```

pshipton commented:
A vmfarm build didn't work either (/build_info.php?build_id=23259), although the problem doesn't seem obviously related to these changes. I'm not sure yet what else it could be.

JasonFengJ9 commented:
> A vmfarm build didn't work either (/build_info.php?build_id=23259)

```
    [javac] error: Exception thrown while constructing Processor object: JVMCFRE003 bad major version; class=org/apache/logging/log4j/core/config/plugins/processor/PluginProcessor, offset=6
  [antcall] Exiting /bluebird/builds/bld_23259/testing/com.ibm.j9.jit.fiorano/build.xml.
   [subant] Exiting /bluebird/builds/bld_23259/testing/com.ibm.j9.jit.fiorano/build.xml.
```

pshipton commented:
The vmfarm build problem also occurs without these changes so something else has caused it. I'll try these changes again once the other problem is resolved.

pshipton commented:
A new vmfarm build passed.

pshipton commented:
Although that was before the latest changes. I'll run a new one.

pshipton commented:
Latest changes also passed on vmfarm.

pshipton commented:
jenkins copyright check

pshipton commented:
jenkins line endings check

pshipton commented:
lgtm, I can see all the new tests running.

JCL_Test_SM
JCL_Test_SM_none_SCC
- JCL_TEST_Java-Lang_ClassLoader_SM
- JCL_TEST_Java-Lang_SM
- JCL_TEST_Java-Lang-Invoke_SM
- JCL_TEST_IBM-VM_SM

JCL_TEST_Java-Security_SM
- JCL_TEST_Java-Security_SM

J9VMInternals_Test_SM
- JCL_TEST_Java-Internals_SM

JasonFengJ9 reviewed (approved):
LGTM
Thanks @EricYangIBM

pshipton commented:
@renfeiw @llxia any further comments?

renfeiw reviewed (approved):
LGTM

llxia reviewed (approved):
LGTM
